scripts/deep_learning/hierarchy/LaterTier/TrainLaterTierModel_showers.ipynb

```diff
@@ -0,0 +1,897 @@
+    "    classifier_model.train()\n",
+    "    \n",
+    "    # Iterate in batches over the training dataset.                        \n",
+    "    for features_0, features_1, target_0, target_1, target, trueGen in loader_train:  \n",
+    "        \n",
+    "        # Skip incomplete batches\n",
+    "        if (target.shape[0] != BATCH_SIZE) :\n",
+    "            continue        \n",
+    "            \n",
+    "        # Get predictions\n",
+    "        pred_0 = model_branch(features_0)\n",
+    "        pred_1 = model_branch(features_1)\n",
+    "        classifier_pred = classifier_model(torch.concatenate((pred_0, pred_1), axis=1))\n",
+    "        \n",
+    "        # Get loss\n",
+    "        target_0 = target_0.to(torch.long)\n",
+    "        target_1 = target_1.to(torch.long)\n",
+    "        classifier_target = target.to(torch.long).reshape(-1,1)\n",
+    "        loss_0 = loss_function_branch(pred_0, target_0, trueGen, classWeights_branch)        \n",
+    "        loss_1 = loss_function_branch(pred_1, target_1, trueGen, classWeights_branch) \n",
+    "        classifier_loss = loss_function_classifier(classifier_pred, classifier_target, trueGen, classWeights_classifier)\n",
+    "        total_loss = loss_0 + loss_1 + classifier_loss\n",
+    "        \n",
+    "        # Update model parameters\n",
+    "        optimiser.zero_grad()\n",
+    "        total_loss.backward()\n",
+    "        optimiser.step()\n",
+    "        \n",
+    "    ######################\n",
+    "    # Validation metrics \n",
+    "    ######################\n",
+    "    with torch.no_grad():\n",
+    "        \n",
+    "        # Begin testing mode\n",
+    "        model_branch.eval()\n",
+    "        classifier_model.eval()\n",
+    "        \n",
+    "        # Initialise metrics        \n",
+    "        total_batches_train = 0        \n",
+    "        total_loss_0_train = 0\n",
+    "        true_scores_0_train = []\n",
+    "        background_scores_0_train = []\n",
+    "        wrong_orientation_scores_0_train = []\n",
+    "        total_loss_1_train = 0\n",
+    "        true_scores_1_train = []\n",
+    "        background_scores_1_train = []\n",
+    "        wrong_orientation_scores_1_train = []\n",
+    "        total_classification_loss_train = 0\n",
+    "        positive_scores_train = []\n",
+    "        negative_scores_train = []\n",
+    "        \n",
+    "        total_batches_test = 0        \n",
+    "        total_loss_0_test = 0\n",
+    "        true_scores_0_test = []\n",
+    "        background_scores_0_test = []\n",
+    "        wrong_orientation_scores_0_test = []\n",
+    "        total_loss_1_test = 0\n",
+    "        true_scores_1_test = []\n",
+    "        background_scores_1_test = []\n",
+    "        wrong_orientation_scores_1_test = []\n",
+    "        total_classification_loss_test = 0\n",
+    "        positive_scores_test = []\n",
+    "        negative_scores_test = []        \n",
```

Comment:
As above, can be refactored to some metrics class with train and test instances.

Reply:
yep, done